### PR TITLE
Support for partial indices

### PR DESCRIPTION
**Pull request by dyemanov**
As requested in #3750. Also supported by MS SQL, PostgreSQL, SQLite, MongoDB. Brief description can be found in [Wikipedia](https://en.wikipedia.org/wiki/Partial_index).

### CODE REVIEW

**Comment by livius2** [comment]
Thank you for implementing this. Next step partial `Foreign Key` like in my comment in https://github.com/FirebirdSQL/firebird/issues/3750 :)

File: src/burp/backup.epp [hunk 1]
```diff
@@ -1617,68 +1617,152 @@ void put_index( burp_rel* relation)
 	// requests--this requires more code but it is well worth it
 	// for the performance benefits, especially remotely--deej
 
-	FOR (REQUEST_HANDLE tdgbl->handles_put_index_req_handle1)
-		X IN RDB$INDICES WITH
-		X.RDB$RELATION_NAME EQ relation->rel_name
+	if (tdgbl->runtimeODS >= DB_VERSION_DDL13_1)
```
Comment (asfernandes) [inline src/burp/backup.epp:1620]: Diff comparation is not easy here.

Can't we use unique code where non-existent fields are read as NULL?

Or are you preventing problem of new gbak being connected to very old engine that treat non-existent fields as error?

File: src/burp/restore.epp [hunk 2]
```diff
@@ -6616,6 +6649,40 @@ bool get_index(BurpGlobals* tdgbl, const burp_rel* relation)
 		general_on_error ();
 	END_ERROR;
 
+	if (hasConditionSrc || hasConditionBlr)
```
Comment (asfernandes) [inline src/burp/restore.epp:6652]: Same here. Write to non-existent fields is possible since many versions ago.

**Comment by asfernandes** [comment]
I suppose that if v4 opens an ODS 13.1 database with partial index, it may not work properly, trying to find entries in index without considering its condition, correct?

Do we consider these cases as valid for minor ODS upgrade?

**Comment by dyemanov** [comment]
> I suppose that if v4 opens an ODS 13.1 database with partial index, it may not work properly, trying to find entries in index without considering its condition, correct?
> 
> Do we consider these cases as valid for minor ODS upgrade?

v4 cannot attach an ODS 13.1 database.

File: src/burp/backup.epp (hunk 1 above)
Comment (dyemanov) [reply src/burp/backup.epp:1620]: Yes, this is to please old engine versions.

**Comment by pavel-zotov** [comment]
Following script causes FB crash (checked on 5.0.0.938; problem in line marked as "[ ! ]"):
```
set bail on;
shell del c:\temp\tmp4test.fdb 2>nul;
create database 'localhost:c:\temp\tmp4test.fdb';
show version;
set list on;
set echo on;
recreate table test(
    id int primary key
    ,f01 int
    ,f02 int
);
create index test_f01 on test computed by (f01)
where f01 = 1 --------------------------------- [ ! ]
;
```

**Comment by dyemanov** [comment]
@pavel-zotov Should be fixed now.

**Comment by pavel-zotov** [comment]
@@@ QA issue @@@
Currently this test contains only trivial cases for check.
More complex examples, including test of backup/restore, validation, using of misc datatypes (non-ascii, decfloat and int128), will be added later.

**Pull request merged**